From a505c204dd7ba3a5abb2485b685b8190a5bf31f7 Mon Sep 17 00:00:00 2001
From: Jiri Olsa <jolsa@redhat.com>
Date: Mon, 25 Aug 2014 07:53:01 -0400
Subject: [tools] perf/symbols: Fix JIT symbol resolution on heap

Message-id: <1408953210-26343-294-git-send-email-jolsa@redhat.com>
Patchwork-id: 89036
O-Subject: [PATCH RHEL7.1 BZ1133083 293/322] perf symbols: Fix JIT symbol resolution on heap
Bugzilla: 1133083
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>

From: Jiri Olsa <jolsa@kernel.org>

Bugzilla: 1133083
https://bugzilla.redhat.com/show_bug.cgi?id=1133083

upstream
========
commit 578c03c86fadcc6fd7319ddf41dd4d1d88aab77a
Author: Namhyung Kim <namhyung@kernel.org>
Date: Thu Jan 16 10:49:31 2014 +0900

description
===========
Gaurav reported that perf cannot profile JIT program if it executes the
code on heap.  This was because current map__new() only handle JIT on
anon mappings - extends it to handle no_dso (heap, stack) case too.

This patch assumes JIT profiling only provides dynamic function symbols
so check the mapping type to distinguish the case.  It'd provide no
symbols for data mapping - if we need to support symbols on data
mappings later it should be changed.
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/tools/perf/util/map.c b/tools/perf/util/map.c
index 9b9bd71..ee1dd68 100644
--- a/tools/perf/util/map.c
+++ b/tools/perf/util/map.c
@@ -69,7 +69,7 @@ struct map *map__new(struct list_head *dsos__list, u64 start, u64 len,
   map->ino = ino;
   map->ino_generation = ino_gen;
 
-  if (anon) {
+  if ((anon || no_dso) && type == MAP__FUNCTION) {
    snprintf(newfilename, sizeof(newfilename), "/tmp/perf-%d.map", pid);
    filename = newfilename;
   }
@@ -93,7 +93,7 @@ struct map *map__new(struct list_head *dsos__list, u64 start, u64 len,
     * functions still return NULL, and we avoid the
     * unnecessary map__load warning.
     */
-   if (no_dso)
+   if (type != MAP__FUNCTION)
     dso__set_loaded(dso, map->type);
   }
  }
-- 
1.7.1